### PR TITLE
Fix EOF detection in LP file reader

### PR DESCRIPTION
**Pull request by fuglede**
The condition `this->linebufferpos == this->linebuffer.size()` would never be `true` except potentially (but not necessarily) at the end of the file, since whenever an `'\n'` is encountered, `this->linebufferpos` is manually set to correspond to the end of the line (on line 998).

Moreover, the condition would mean that an actual EOF would not be properly handled when the final line of the file is not empty, which would instead cause an infinite loop as the final line is read over and over. In particular, this infinite loop is encountered for the example given in gh-840:

```
4d 69 6e 69 6d 69 7a 65 0a 20 78 30 20 2b 20 78 31 0a 73 75 62 6a 65 63 74 20 74 6f 0a 20 63 31 3a 20 78 30 20 2b 20 78 31 20 3c 3d 20 31 0a 45 6e 64
```

Here we just move the EOF check to before reading a new line.

This closes gh-840.

Now at the same time, it certainly seems like the condition wouldn't be put there for no reason, so maybe I'm missing something.

### CODE REVIEW

File: extern/filereaderlp/reader.cpp [hunk 1]
```diff
@@ -903,8 +903,8 @@ void Reader::readnexttoken(bool& done) {
      this->linebufferpos = 0;
    }
 
-   // if all line has been read and we are at end of file, then stop
-   if (this->linebufferpos == this->linebuffer.size() && this->file.eof()) {
```
Comment (fuglede) [inline extern/filereaderlp/reader.cpp:907]: So, one case this doesn't handle is what happens if "End" is not on its own line. I guess that's no good.

Edit: Fixed by simply making the check before reading another line.

File: extern/filereaderlp/reader.cpp [hunk 2]
```diff
@@ -892,7 +892,12 @@ void Reader::readnexttoken(bool& done) {
    done = false;
 
    if (this->linebufferpos == this->linebuffer.size()) {
-     // read next line
+     // read next line if any are left. 
+     if (this->file.eof()) {
+         this->rawtokens.push_back(std::unique_ptr<RawToken>(new RawToken(RawTokenType::FLEND)));
+         done = true;
```
Comment (fuglede) [inline extern/filereaderlp/reader.cpp:898]: The value of `done` is never actually used for anything; I can remove that in a separate PR if relevant.

Comment (jajhall) [reply extern/filereaderlp/reader.cpp:898]: Thanks, but let's leave that to @feldmeier , who might want it left in for future use

**Review by svigerske — APPROVED** [review]
Seems like I messed this up.
I didn't think that `std::getline(this->file, linebuffer);` would leave `linebuffer` untouched when at eof.    
So an alternative fix would be
```diff
--- a/extern/filereaderlp/reader.cpp
+++ b/extern/filereaderlp/reader.cpp
@@ -893,6 +893,7 @@ void Reader::readnexttoken(bool& done) {
 
    if (this->linebufferpos == this->linebuffer.size()) {
      // read next line
+     linebuffer.clear();
      std::getline(this->file, linebuffer);
 
      // drop \r
```
But moving the eof check is also good (maybe also better).

Adding `fail.lp` from #840 to the unittests would be nice.

**Comment by fuglede** [comment]
> Adding fail.lp from https://github.com/ERGO-Code/HiGHS/issues/840 to the unittests would be nice.

I've added such a test to `TestFilereader.cpp`.

**Comment by jajhall** [comment]
> So an alternative fix would be
> 
> ```diff
> --- a/extern/filereaderlp/reader.cpp
> +++ b/extern/filereaderlp/reader.cpp
> @@ -893,6 +893,7 @@ void Reader::readnexttoken(bool& done) {
>  
>     if (this->linebufferpos == this->linebuffer.size()) {
>       // read next line
> +     linebuffer.clear();
>       std::getline(this->file, linebuffer);
>  
>       // drop \r
> ```
> 
> But moving the eof check is also good (maybe also better).

What do you reckon @fuglede ?

**Comment by fuglede** [comment]
> What do you reckon @fuglede ?

I'm probably biased since this solution is the one that happened to be what I found natural first. Perhaps one can argue that sticking to `eof` makes semantics clearer?

Another thing that could also work is

```cpp
     std::getline(this->file, linebuffer);
     if (!this->file)
     {
         this->rawtokens.push_back(std::unique_ptr<RawToken>(new RawToken(RawTokenType::FLEND)));
         done = true;
         return;
     }
```

File: extern/filereaderlp/reader.cpp [hunk 3]
```diff
@@ -892,9 +892,14 @@ void Reader::readnexttoken(bool& done) {
    done = false;
 
    if (this->linebufferpos == this->linebuffer.size()) {
-     // read next line
+     // read next line if any are left. 
      std::getline(this->file, linebuffer);
-
+     if (this->file.eof())
```
Comment (svigerske) [inline extern/filereaderlp/reader.cpp]: Doesn't this now skip parsing the content of the last line, if not empty?

I was also ok with the previous solution.

Comment (fuglede) [reply extern/filereaderlp/reader.cpp]: You're right; `eof` triggers on the last line, even if it has contents. I guess it's safe anyway given that the end marker appears on a line of it's own, but relying on that seems iffy. Let me just revert the last commit.

**Comment by feldmeier** [comment]
thanks for fixing this, changes look fine to me

**Pull request merged**